Proposal: slo.yaml: Use filters as placeholders for sli-providers

christian-kreuzberger-dtx opened the pull request:
Within this proposal we change the meaning of `filters` within slo.yaml.

Key/Value pairs defined within the filters section are already partially used as placeholders within the dynatrace-sli-service and prometheus-sli-service that are currently being developed.

For example,
```yaml
filter:
   handler: "ItemsController.addToCart"
```

could be used together with the following SLI configuration:
```yaml
apiVersion: v1
data:
  custom-queries: |
    cpu_usage: avg(rate(container_cpu_usage_seconds_total{namespace="$PROJECT-$STAGE",pod_name=~"$SERVICE-primary-.*"}[5m]))
    response_time_p95: histogram_quantile(0.95, sum by(le) (rate(http_response_time_milliseconds_bucket{handler="$HANDLER",job="$SERVICE-$PROJECT-$STAGE-canary"}[$DURATION_SECONDS])))
kind: ConfigMap
metadata:
  name: prometheus-sli-service-config-sockshop
  namespace: keptn
```

edit: Furthermore, this change would allow us to leave filters empty by default.

dirkwall commented:
I propose to rename `filters` to `scope` and maybe provide an example that doesn't have `handler=$HANDLER` but different names for key and value, to avoid ambiguity.

christian-kreuzberger-dtx commented:
@dirkwall I have renamed filter to scope, change `handler:` to `endpoint:` to not be ambiguous. 
I've also added the prometheus-sli config example into the spec, to make clear how a scope can be used.

dirkwall commented:
Can we please make it more obvious, that the `$ENDPOINT` in the SLI provider ConfigMap refers to the `endpoint` parameter from the `scope` section of the `slo.yaml`?
When reading the spec as it is right now, that is implicit knowledge and I'd like to point that out explicitly.

christian-kreuzberger-dtx commented:
I was trying to avoid mentioning *prometheus-sli-provider* within the spec in the first place. It was not present in the [first commit](https://github.com/keptn/spec/pull/7/commits/4f3c3001dcd94ac9fbdad2faf18761b04d5cebbc) for this PR at all, but in the description of this PR to explain the idea behind the change.

The fact that it is available as a variable/placeholder is an implementation detail of the SLI provider, and not the spec.

Where do we draw the line in the spec between "this is what is specified" vs. "this is how it is used"?

christian-kreuzberger-dtx commented:
Not relevant right now, closing
